ambry-router/src/main/java/com/github/ambry/router/GetBlobOperation.java

```diff
@@ -1353,12 +1356,19 @@ private void handleMetadataBlob(BlobData blobData, byte[] userMetadata, ByteBuff
                 new DecryptJob(blobId, encryptionKey, null, ByteBuffer.wrap(userMetadata), cryptoService, kms,
                     decryptJobMetricsTracker, (DecryptJob.DecryptJobResult result, Exception exception) -> {
                   routerMetrics.decryptTimeMs.update(System.currentTimeMillis() - startTimeMs);
-                  decryptJobMetricsTracker.onJobCallbackProcessingStart();
-                  logger.trace("Handling decrypt job call back for Metadata chunk {} to set decrypt callback results",
-                      blobId);
-                  decryptCallbackResultInfo.setResultAndException(result, exception);
-                  routerCallback.onPollReady();
-                  decryptJobMetricsTracker.onJobCallbackProcessingComplete();
+                  if (isOperationComplete() || operationException.get() != null) {
+                    if (result != null && result.getDecryptedBlobContent() != null) {
+                      result.getDecryptedBlobContent().release();
+                    }
+                  } else {
+                    decryptJobMetricsTracker.onJobCallbackProcessingStart();
+                    logger.trace("Handling decrypt job call back for Metadata chunk {} to set decrypt callback results",
+                        blobId);
+                    decryptCallbackResultInfo.setResultAndException(result, exception);
+                    routerCallback.onPollReady();
+                    decryptJobMetricsTracker.onJobCallbackProcessingComplete();
+                    processCallbacks();
```

Comment:
i'm a little worried about calling processCallbacks from a different thread in a callback. I don't think it was designed to be called from non event loop threads. Will take a deeper look to see if there's any alternative (or if it is already thread safe)

Reply:
you are right, the progresstracker and someother object is not thread safe and should probably only be used in event loop thread. I will update it.